src/mapped_tcp_socket.rs

```diff
@@ -36,18 +50,376 @@ pub struct MappedTcpSocket {
+                                                      local_addr_v4, 0,
+                                                      "rust nat_traversal")
+                        {
+                            Ok(external_addr) => {
+                                endpoints.push(MappedSocketAddr {
+                                    addr: SocketAddr(net::SocketAddr::V4(external_addr)),
+                                    nat_restricted: false,
+                                });
+                            },
+                            Err(e) => {
+                                warnings.push(MappedTcpSocketMapWarning::GetExternalPort {
+                                    gateway_addr: gateway.addr,
+                                    err: e,
+                                });
+                            }
+                        }
+                    };
+                };
+            },
+            IpAddr::V6(ipv6_addr) => {
+                if socket_utils::ipv6_is_unspecified(&ipv6_addr) {
+                    // If the socket address is unspecified add an address for every interface.
+                    for iface_v6 in mapping_context::interfaces_v6(&mc) {
+                        let local_iface_addr = net::SocketAddr::V6(net::SocketAddrV6::new(iface_v6.addr, local_addr.port(), 0, 0));
+                        endpoints.push(MappedSocketAddr {
+                            addr: SocketAddr(local_iface_addr),
+                            nat_restricted: false,
+                        });
+                    };
+                }
+                else {
+                    endpoints.push(MappedSocketAddr {
+                        addr: SocketAddr(net::SocketAddr::V6(net::SocketAddrV6::new(ipv6_addr, local_addr.port(), 0, 0))),
+                        nat_restricted: false,
+                    });
+                }
+            },
+        };
+        
+        let mut mapping_threads = Vec::new();
+        let simple_servers = mapping_context::simple_tcp_servers(&mc);
+        for simple_server in simple_servers {
+            mapping_threads.push(thread::spawn(move || {
+                let mapping_socket = match new_reusably_bound_socket(&local_addr) {
+                    Ok(mapping_socket) => mapping_socket,
+                    Err(e) => return Err(MappedTcpSocketMapWarning::NewReusablyBoundSocket { err: e }),
+                };
+                let mut stream = match mapping_socket.connect(&*simple_server) {
+                    Ok(stream) => stream,
+                    Err(e) => return Err(MappedTcpSocketMapWarning::MappingSocketConnect {
+                        addr: simple_server,
+                        err: e
+                    }),
+                };
+                let send_data = listener_message::REQUEST_MAGIC_CONSTANT;
+                // TODO(canndrew): What should we do if we get a partial write?
+                let _ = match stream.write(&send_data[..]) {
```

Comment:
You can use [`write_all`](https://doc.rust-lang.org/std/io/trait.Write.html#method.write_all).